File: tests/validate_cli.js
#!/usr/bin/env node
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, rmSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = dirname(dirname(fileURLToPath(import.meta.url)));

function readJson(path) {
  return JSON.parse(readFileSync(path, "utf8"));
}

function runCli(args) {
  return spawnSync(process.execPath, [join(root, "bin/ec-image.js"), ...args], {
    cwd: root,
    encoding: "utf8",
  });
}

function validatePackageMetadata() {
  const packagePath = join(root, "package.json");
  assert.equal(existsSync(packagePath), true, "package.json should exist for npm usage");

  const pkg = readJson(packagePath);
  assert.equal(pkg.type, "module", "package should use ESM for the Node CLI");
  assert.equal(pkg.bin?.["ec-image"], "./bin/ec-image.js", "package should expose ec-image bin");
  assert.equal(pkg.scripts?.test, "node tests/validate_cli.js && node tests/validate_plugin.js");
}

function validateMainPlanGeneration() {
  const outDir = join(root, "tmp/cli-test-main");
  rmSync(outDir, { recursive: true, force: true });
  mkdirSync(outDir, { recursive: true });

  const result = runCli([
    "main",
    "--input",
    "public/demo-results/main/watch-main-1.png",
    "--product-name",
    "男士机械腕表",
    "--category",
    "watch",
    "--platform",
    "taobao",
    "--count",
    "5",
    "--mode",
    "strict",
    "--out",
    outDir,
  ]);

  assert.equal(
    result.status,
    0,
    `CLI should succeed\nSTDOUT:\n${result.stdout}\nSTDERR:\n${result.stderr}`,
  );
  for (const expectedLine of [
    "[1/6] Parsed request",
    "[2/6] Checked reference image",
    "[3/6] Planned image types",
    "[4/6] Built planning package",
    "[5/6] Wrote files",
    "[6/6] Ready for Codex or image generation",
    "status: found",
    "ok brief.json",
    "ok main-image-plan.json",
    "ok prompts.json",
    "ok qa-checklist.json",
    "ok summary.md",
    "output:",
    "next: review summary.md",
    "brief.json",
    "main-image-plan.json",
    "prompts.json",
    "qa-checklist.json",
    "summary.md",
  ]) {
    assert.match(result.stdout, new RegExp(expectedLine.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
  }

  for (const name of [
    "brief.json",
    "main-image-plan.json",
    "prompts.json",
    "qa-checklist.json",
    "summary.md",
  ]) {
    assert.equal(existsSync(join(outDir, name)), true, `CLI should create ${name}`);
  }

  const brief = readJson(join(outDir, "brief.json"));
  assert.equal(brief.product_name, "男士机械腕表");
  assert.equal(brief.mode, "strict");

  const plan = readJson(join(outDir, "main-image-plan.json"));
  assert.deepEqual(
    plan.images.map((item) => item.image_type),
    ["white-background", "premium-studio", "lifestyle", "comparison", "social-cover"],
  );

  const prompts = readJson(join(outDir, "prompts.json"));
  assert.equal(prompts.prompts.length, 5);
  assert.match(prompts.prompts[0].reference_instruction, /preserve product/);

  const qa = readJson(join(outDir, "qa-checklist.json"));
  assert.match(qa.checks.join("\n"), /product remains the main subject/);
  assert.match(qa.checks.join("\n"), /white-background/);
  assert.match(qa.checks.join("\n"), /watermarks/);

  const summary = readFileSync(join(outDir, "summary.md"), "utf8");
  assert.match(summary, /男士机械腕表/);
  assert.match(summary, /white-background/);
  assert.match(summary, /prompts\.json/);
}

function validateErrorHandling() {
  const result = runCli([
    "main",
    "--input",
    "public/demo-results/main/missing.png",
    "--out",
    join(root, "tmp/cli-test-missing"),
  ]);

  assert.equal(result.status, 2, "missing input should exit with code 2");
  assert.match(result.stderr, /input image does not exist/);
}

validatePackageMetadata();
validateMainPlanGeneration();
validateErrorHandling();
console.log("cli validation passed");

File: tests/validate_cli.py
import json
import subprocess
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


def read_json(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def run_cli(args: list[str]) -> subprocess.CompletedProcess:
    return subprocess.run(
        [str(ROOT / "bin/ec-image"), *args],
        cwd=ROOT,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )


def validate_main_plan_generation() -> None:
    out_dir = ROOT / "tmp/cli-test-main"
    if out_dir.exists():
        for child in out_dir.iterdir():
            child.unlink()
    else:
        out_dir.mkdir(parents=True)

    result = run_cli(
        [
            "main",
            "--input",
            "public/demo-results/main/watch-main-1.png",
            "--product-name",
            "男士机械腕表",
            "--category",
            "watch",
            "--platform",
            "taobao",
            "--count",
            "5",
            "--mode",
            "strict",
            "--out",
            str(out_dir),
        ]
    )
    if result.returncode != 0:
        raise AssertionError(f"CLI failed:\nSTDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}")

    expected_files = [
        "brief.json",
        "main-image-plan.json",
        "prompts.json",
        "qa-checklist.json",
        "summary.md",
    ]
    missing = [name for name in expected_files if not (out_dir / name).exists()]
    if missing:
        raise AssertionError(f"CLI did not create expected files: {missing}")

    brief = read_json(out_dir / "brief.json")
    if brief["product_name"] != "男士机械腕表":
        raise AssertionError("brief.json should preserve product_name")
    if brief["mode"] != "strict":
        raise AssertionError("brief.json should preserve mode")

    plan = read_json(out_dir / "main-image-plan.json")
    image_types = [item["image_type"] for item in plan["images"]]
    if image_types != [
        "white-background",
        "premium-studio",
        "lifestyle",
        "comparison",
        "social-cover",
    ]:
        raise AssertionError(f"unexpected main image type order: {image_types}")

    prompts = read_json(out_dir / "prompts.json")
    if len(prompts["prompts"]) != 5:
        raise AssertionError("prompts.json should include 5 prompt packs")
    if "preserve product" not in prompts["prompts"][0]["reference_instruction"]:
        raise AssertionError("prompt pack should include reference preservation instruction")

    qa = read_json(out_dir / "qa-checklist.json")
    checks = "\n".join(qa["checks"])
    for expected in ["product remains the main subject", "white-background", "watermarks"]:
        if expected not in checks:
            raise AssertionError(f"qa-checklist.json is missing QA text: {expected}")

    summary = (out_dir / "summary.md").read_text(encoding="utf-8")
    for expected in ["男士机械腕表", "white-background", "prompts.json"]:
        if expected not in summary:
            raise AssertionError(f"summary.md is missing expected text: {expected}")


def main() -> None:
    validate_main_plan_generation()
    print("cli validation passed")


if __name__ == "__main__":
    main()
